Increase limit for float vectors to 2048

mkleen opened the pull request:
## Summary of the changes / Why this improves CrateDB

This increases the limit of float vector from 1024 to 2048.
The previous limit was based on what Lucene provided but current
discussions and current benchmarks indicate that 2048 will also
be ok and the next Lucene version will have 2048 as default.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mkleen commented:
retest this please

mkleen commented on server/src/main/java/org/elasticsearch/index/mapper/FloatVectorFieldMapper.java:
```diff
@@ -46,7 +49,56 @@ public static final class Defaults {
-        public static final FieldType FIELD_TYPE = new FieldType();
+        public static final int DIMENSION_LIMIT = 2048;
+
+        public static final FieldType FIELD_TYPE = new FieldType() {
```
I know it looks ugly, the reason for this is that dimension is private and cannot be accessed from a subclass. The only alternative I came up with is to re-implement the whole `FieldType` which would duplicate all the code of FieldType.

mkleen commented:
retest this please

mkleen commented:
retest this please

mfussenegger commented on server/src/test/java/io/crate/integrationtests/KnnIntegrationTest.java:
```diff
@@ -0,0 +1,53 @@
+    // does only work with http, but not with postgres protocol
+    @UseJdbc(0)
+    public void test_insert_select_and_match_2048_float_vector() {
+        int size = 2048;
+        execute("create table tbl(x float_vector(" + size + "));");
+        var values = new ArrayList<String>();
+        for (int i = 0; i < size; i++) {
+            values.add(Float.toString(randomFloat()));
+        }
+        var valuesAsString = "[" + String.join(",", values) + "]";
+        // does not work as an argument for a prepared statement
+        execute("insert into tbl(x) values (" + valuesAsString + ")");
```
```suggestion
        execute("create table tbl(x float_vector(2048));");
        float[] vector = new float[2048];
        for (int i = 0; i < vector.length; i++) {
            vector[i] = randomFloat();
        }
        execute("insert into tbl(x) values (?)", new Object[] { vector });
```

The string conversions shouldn't be necessary and using float[] has a lower memory footprint.

mfussenegger commented on server/src/test/java/io/crate/integrationtests/KnnIntegrationTest.java:
```diff
@@ -0,0 +1,53 @@
+        execute("refresh table tbl");
+        execute("select count(*) from tbl");
+        assertThat(response).hasRowCount(1);
+        execute("SELECT x, _score FROM tbl WHERE knn_match(x, " + valuesAsString + ", 2)");
```
```suggestion
        execute("SELECT x, _score FROM tbl WHERE knn_match(x, ?, 2)", new Object[] { vector });
```

mfussenegger reviewed (commented):
Would be nice to have a rationale in the commit message that explains why we think it's a good idea to raise a limit set by Lucene (which they hopefully did set for a good reason too?)

mfussenegger commented on server/src/test/java/io/crate/integrationtests/KnnIntegrationTest.java:
```diff
@@ -0,0 +1,53 @@
+
+public class KnnIntegrationTest extends IntegTestCase {
+
+    // does only work with http, but not with postgres protocol
```
```suggestion
    // Requires https://github.com/crate/crate/pull/14712 to work with JDBC
```

mkleen commented:
It looks like the Lucene community actually close to agree on a 2048 limit 
https://github.com/apache/lucene/issues/11507#issuecomment-1585637725
Especially with the improvements of project panama which look promising
https://github.com/apache/lucene/issues/11507#issuecomment-1607892454

mfussenegger commented:
The limit is also mentioned in https://github.com/crate/crate/blob/master/docs/general/ddl/data-types.rst which would need an update. (And it would be really cool if we could also mention scaling behavior with larger dimensions if there is data about that, e.g. does indexing cost increase linear/exponential/other, and how does it affect storage space)

mkleen commented:
I added the feedback and added a rationale in the commit message. Regarding scaling etc., this will need some more research probably by looking into other projects that use the Lucene vector support. I will do this as a follow-up.

matriv reviewed (approved):
LGTM, thx

mfussenegger commented on server/src/test/java/io/crate/integrationtests/KnnIntegrationTest.java:
```diff
@@ -0,0 +1,50 @@
+
+public class KnnIntegrationTest extends IntegTestCase {
+
+    // Requires https://github.com/crate/crate/pull/14712 to work with JDBC
+    @UseJdbc(0)
+    public void test_insert_select_and_match_2048_float_vector() {
+        execute("create table tbl(x float_vector(2048));");
```
On second thought - do we need a full blown integration test for this, or would it be possible to create a test case in `FloatVectorIndexerTest`?

mkleen replied:
This test caught the issues with JDBC. I can move it to `FloatVectorIndexerTest` but i think one basic integration test for simple use-case is ok.

mkleen replied:
It's converted to FloatVectorIndexerTest.

mfussenegger reviewed (approved):
Would be nice if we could replace the itest with a cheaper version. Otherwise lgtm
